compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs

```diff
@@ -0,0 +1,207 @@
+impl VtableCtx {
+    /// Add a possible implementation for a virtual method call.
+    pub fn add_possible_method(
+        &mut self,
+        trait_name: InternedString,
+        method: usize,
+        imp: InternedString,
+    ) {
+        assert!(self.emit_vtable_restrictions);
+        let key = TraitDefinedMethod { trait_name: trait_name.clone(), vtable_idx: method };
+
+        if let Some(possibilities) = self.possible_methods.get_mut(&key) {
+            possibilities.push(imp);
+        } else {
+            self.possible_methods.insert(key, vec![imp]);
+        }
+    }
+
+    /// The vtable index for drop
+    pub fn drop_index() -> usize {
+        rustc_middle::ty::COMMON_VTABLE_ENTRIES_DROPINPLACE
+    }
+}
+
+/// Internal tracking helpers
+impl VtableCtx {
+    fn get_call_site_global_idx(&mut self) -> usize {
+        assert!(self.emit_vtable_restrictions);
+        self.call_site_global_idx += 1;
+        self.call_site_global_idx
+    }
+
+    /// Add a given call site for a virtual function, incremementing the call
+    /// site index.
+    fn add_call_site(
+        &mut self,
+        trait_name: InternedString,
+        method: usize,
+        function_name: InternedString,
+    ) {
+        assert!(self.emit_vtable_restrictions);
+        let site = CallSite {
+            trait_method: TraitDefinedMethod { trait_name: trait_name, vtable_idx: method },
+            function_name: function_name,
+        };
+        self.call_sites.push(site);
+    }
+}
+
+impl<'tcx> GotocCtx<'tcx> {
+    /// Wrap a virtual call through a function pointer and restrict the
+    /// possible targets.
+    ///
+    /// We need to wrap because for the current implemention, CBMC employs
+    /// a hard-to-get-right naming scheme for restrictions: the call site is
+    /// named for its index in  a given function. We don't have a good way to
+    /// track _all_ function pointers within the function, so wrapping the call
+    /// to a function that makes a single virtual function pointer call makes
+    /// the naming unambiguous.
+    ///
+    /// This can be simplified if CBMC implemented label-based restrictions.
+    /// RMC tracking: https://github.com/model-checking/rmc/issues/651
+    /// CBMC tracking: https://github.com/diffblue/cbmc/issues/6464
+    pub fn virtual_call_with_restricted_fn_ptr(
+        &mut self,
+        trait_ty: Type,
+        vtable_idx: usize,
+        fn_ptr: Expr,
+        args: Vec<Expr>,
+    ) -> Expr {
+        assert!(self.vtable_ctx.emit_vtable_restrictions);
+
+        // Crate-based naming scheme for wrappers
+        let full_crate_name = self.full_crate_name().to_string().replace("::", "_");
+        let wrapper_name: InternedString = format!(
+            "restricted_call_{}_{}",
+            full_crate_name,
+            self.vtable_ctx.get_call_site_global_idx()
+        )
+        .into();
+
+        // We only have the Gotoc type, we need to normalize to match the MIR type.
```

Comment:
Is there any risk here if MIR changes?

Reply:
Less so now that we have the `type_map`, we are normalizing the MIR on both sides instead of relying on string editing.